spawn_worker/src/unix/spawn.rs

```diff
@@ -48,61 +64,155 @@ impl ExecVec {
+    #[cfg(target_os = "macos")]
+    pub fn detect_spawn_method(&self) -> std::io::Result<SpawnMethod> {
+        Ok(SpawnMethod::Exec)
     }
 
-    #[cfg(not(target_os = "linux"))]
-    fn default() -> Self {
-        Self::Exec
+    /// Automatically detect which spawn method should be used
+    #[cfg(not(target_os = "macos"))]
+    pub fn detect_spawn_method(&self) -> std::io::Result<SpawnMethod> {
+        use std::path::PathBuf;
+        let current_exec_path = env::current_exe()?;
+        let current_exec_filename = current_exec_path.file_name().unwrap_or_default();
+        #[cfg(target_os = "linux")]
+        let default_method = SpawnMethod::FdExec;
+
+        #[cfg(not(target_os = "linux"))]
+        let default_method = SpawnMethod::Exec;
+
+        let target_path = match self {
+            Target::Entrypoint(e) => e.get_fs_path().ok_or_else(|| {
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "can't find the entrypoint's target path",
+                )
+            }),
+            Target::Manual(p, _) => p
+                .to_str()
+                .map(PathBuf::from)
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e)),
+            Target::Noop => return Ok(default_method),
+        }?;
+        let target_filename = target_path.file_name().ok_or_else(|| {
+            std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "can't extract actual target filename",
+            )
+        })?;
+
+        // simple heuristic that should cover most cases
+        // if both executable path and target's entrypoint path end up having the same filenames
+        // then it means its not a shared library - and we need to load the trampoline us ld_preload
+        if current_exec_filename == target_filename {
+            Ok(SpawnMethod::LdPreload)
+        } else {
+            Ok(default_method)
+        }
     }
 }
 
-pub enum Target {
-    ViaFnPtr(extern "C" fn()),
-    Manual(CString, CString),
-    Fork(fn()),
-    Noop,
+enum ChildStdio {
+    Inherit,
+    Owned(OwnedFd),
+    Ref(libc::pid_t),
+}
+
+impl ChildStdio {
+    fn as_fd(&self) -> Option<libc::pid_t> {
+        match self {
+            ChildStdio::Inherit => None,
+            ChildStdio::Owned(fd) => Some(fd.as_raw_fd()),
+            ChildStdio::Ref(fd) => Some(*fd),
+        }
+    }
+}
+
+pub enum Stdio {
+    Inherit,
+    Fd(OwnedFd),
+    Null,
+}
+
+impl Stdio {
+    fn as_child_stdio(&self) -> std::io::Result<ChildStdio> {
+        match self {
+            Stdio::Inherit => Ok(ChildStdio::Inherit),
+            Stdio::Fd(fd) => {
+                if fd.as_raw_fd() >= 0 && fd.as_raw_fd() <= libc::STDERR_FILENO {
+                    Ok(ChildStdio::Owned(fd.try_clone()?))
+                } else {
+                    Ok(ChildStdio::Ref(fd.as_raw_fd()))
+                }
```

Comment:
I don't understand this one. Why couldn't we clone non stdin/stdout/stderr file descriptors?

Reply:
this code follows what Command::spawn is doing. Also since stdio fd's are not CLOSEXEC by default then we don't need to dupe them.